sdk/storage/src/blobs/blob_client.cpp

```diff
@@ -222,29 +223,161 @@ namespace Azure { namespace Storage { namespace Blobs {
     BlobDownloadInfo ret = returnTypeConverter(firstChunk);
 
     // Keep downloading the remaining in parallel
-    auto downloadChunkFunc = [&](int64_t offset, int64_t length, int64_t chunkId, int64_t numChunks) {
-
-      DownloadBlobOptions chunkOptions;
-      chunkOptions.Context = options.Context;
-      chunkOptions.Offset = offset;
-      chunkOptions.Length = length;
-      auto chunk = Download(chunkOptions);
-      int64_t bytesRead = Azure::Core::Http::BodyStream::ReadToCount(
-          chunkOptions.Context,
-          *chunk.BodyStream,
-          buffer + (offset - firstChunkOffset),
-          chunkOptions.Length.GetValue());
-      if (bytesRead != chunkOptions.Length.GetValue())
+    auto downloadChunkFunc
+        = [&](int64_t offset, int64_t length, int64_t chunkId, int64_t numChunks) {
+            DownloadBlobOptions chunkOptions;
+            chunkOptions.Context = options.Context;
+            chunkOptions.Offset = offset;
+            chunkOptions.Length = length;
+            auto chunk = Download(chunkOptions);
+            int64_t bytesRead = Azure::Core::Http::BodyStream::ReadToCount(
+                chunkOptions.Context,
+                *chunk.BodyStream,
+                buffer + (offset - firstChunkOffset),
+                chunkOptions.Length.GetValue());
+            if (bytesRead != chunkOptions.Length.GetValue())
+            {
+              throw std::runtime_error("error when reading body stream");
+            }
+
+            if (chunkId == numChunks - 1)
+            {
+              ret = returnTypeConverter(chunk);
+            }
+          };
+
+    int64_t remainingOffset = firstChunkOffset + firstChunkLength;
+    int64_t remainingSize = blobRangeSize - firstChunkLength;
+    int64_t chunkSize;
+    if (options.ChunkSize.HasValue())
+    {
+      chunkSize = options.ChunkSize.GetValue();
+    }
+    else
+    {
+      int64_t c_grainSize = 4 * 1024;
+      chunkSize = remainingSize / options.Concurrency;
+      chunkSize = (std::max(chunkSize, int64_t(1)) + c_grainSize - 1) / c_grainSize * c_grainSize;
+      chunkSize = std::min(chunkSize, c_defaultChunkSize);
+    }
+
+    Details::ConcurrentTransfer(
+        remainingOffset, remainingSize, chunkSize, options.Concurrency, downloadChunkFunc);
+    ret.ContentLength = blobRangeSize;
+    return ret;
+  }
+
+  BlobDownloadInfo BlobClient::DownloadToFile(
+      const std::string& file,
+      const DownloadBlobToFileOptions& options) const
+  {
+    constexpr int64_t c_defaultChunkSize = 8 * 1024 * 1024;
```

Comment:
Why 8MB?

Reply:
I'm fine with any value, but we have to pick a value.

Reply:
I think it's normal to make it 4MB, any other SDK making it 8MB?
Also, I still think it should be the default value of the optional parameter in options, not here.

Reply:
Okay, i'll change it to 4MB.

Reply:
About default value or optional parameter, here's the deal. I want the customer to be able to fully control the block size if they want to, and if they don't, we pick the optimal value for them.

For example, if they're trying to download 4.2MiB blob or blob range with 2 threads. If they explicitly specify the block size is 4MiB, we're going to use 4MiB. If they don't, we'll use 2.1MiB to achieve the best overall performance.